src/unabto/unabto_push.c

```diff
@@ -0,0 +1,344 @@
+#else
+extern uint8_t* unabto_push_notification_get_data(uint8_t* bufStart, const uint8_t* bufEnd, uint32_t seq);
+extern void unabto_push_notification_callback(uint32_t seq, unabto_push_hint* hint);
+
+#endif
+
+/* ---------------------------------------------------- *
+ * Help function definitions                            *
+ * ---------------------------------------------------- */
+void unabto_push_create_and_send_packet(unabto_push_element *elem);
+void unabto_push_set_next_event(void);
+bool unabto_push_verify_integrity(nabto_packet_header* header);
+bool unabto_push_seq_exists(uint32_t seq);
+
+/* ---------------------------------------------------- *
+ * Push global state variables                          *
+ * ---------------------------------------------------- */
+
+unabto_push_element pushSeqQ[NABTO_PUSH_QUEUE_LENGTH];
+/* -------------------------------------------------------*
+ * initialization function to be called before using push *
+ * -------------------------------------------------------*/
+void unabto_push_init(void){
+    NABTO_LOG_TRACE(("BACKOFF PERIOD: %d",UNABTO_PUSH_QUOTA_EXCEEDED_BACKOFF));
+    pushCtx.nextPushEvent = NULL;
+    pushCtx.pushSeqQHead = 0;
+    pushCtx.nextSeq = 0;
+    pushCtx.reattachNeeded = false;
+    pushCtx.lastSent = nabtoGetStamp();
+    pushCtx.backOffLimit = pushCtx.lastSent;
+    size_t i;
+    for (i = 0; i<NABTO_PUSH_QUEUE_LENGTH; i++){
+        pushSeqQ[i].state = UNABTO_PUSH_IDLE;
+        pushSeqQ[i].hint = UNABTO_PUSH_HINT_OK;
+    }
+}
+
+unabto_push_hint unabto_send_push_notification(uint16_t pnsId, uint32_t* seq){
+    nabto_stamp_t now = nabtoGetStamp();
+    // With reattach needed or queue full we must still call callback without releasing Zalgo
+    if(pushCtx.reattachNeeded){
+        pushSeqQ[pushCtx.pushSeqQHead].hint = UNABTO_PUSH_HINT_QUOTA_EXCEEDED_REATTACH;
+    } else if (pushCtx.pushSeqQHead >= NABTO_PUSH_QUEUE_LENGTH){
+        pushSeqQ[pushCtx.pushSeqQHead].hint = UNABTO_PUSH_HINT_QUEUE_FULL;
+    } else {
+        pushSeqQ[pushCtx.pushSeqQHead].hint = UNABTO_PUSH_HINT_OK;
+    }
+    pushSeqQ[pushCtx.pushSeqQHead].seq = pushCtx.nextSeq;
+    *seq = pushCtx.nextSeq;
+    pushCtx.nextSeq++;
+    pushSeqQ[pushCtx.pushSeqQHead].retrans = 0;
+    pushSeqQ[pushCtx.pushSeqQHead].state = UNABTO_PUSH_WAITING_SEND;
+    if(nabtoStampLess(&now, &pushCtx.backOffLimit)){
+        pushSeqQ[pushCtx.pushSeqQHead].stamp = pushCtx.backOffLimit;
+    } else {
+        pushSeqQ[pushCtx.pushSeqQHead].stamp = now;
+    }        
+    pushSeqQ[pushCtx.pushSeqQHead].pnsId = pnsId;
+
+    pushCtx.pushSeqQHead++;
+    unabto_push_set_next_event();
+    return pushSeqQ[pushCtx.pushSeqQHead].hint;
+}
+
+bool unabto_push_notification_remove(uint32_t seq)
+{
+    int i;
+    for(i = 0; i<pushCtx.pushSeqQHead; i++){
+        if(pushSeqQ[i].seq == seq){
+            if(i<NABTO_PUSH_QUEUE_LENGTH-1){
```

Comment:
Change the algorithm to move the top element into the now empty spot, this will simplify the code and make it easier to read. Add a comment which states this is for array compression.